python/pywatchman/__init__.py

```diff
@@ -521,7 +521,60 @@ class ImmutableBserCodec(BserCodec):
         immutable object support """
 
     def _loads(self, response):
-        return bser.loads(response, False)
+        return bser.loads(response, False) # Defaults to BSER v1
+
+
+class Bser2WithFallbackCodec(BserCodec):
+    """ use BSER v2 encoding """
+
+    def __init__(self, transport):
+        super(Bser2WithFallbackCodec, self).__init__(transport)
+        if compat.PYTHON3:
+            self.send(["version", {"required": ["bser-v2"]}])
+        else:
+            self.send(["version", {"optional": ["bser-v2"]}])
+
+        capabilities = self.receive()
+
+        if u'error' in capabilities:
+          raise Exception('Unsupported BSER version')
+
+        if capabilities[u'capabilities'][u'bser-v2']:
+            self.bser_version = 2
+        else:
+            self.bser_version = 1
+
+        assert self.bser_version == 2 # For the sake of testing, let fallback
```

Comment:
Could you elaborate about what you're doing with this? These sorts of assertions should live in tests, not in code.